e2e/test/BulkOperationsE2ETests.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    public class BulkOperationsE2ETests : IDisposable
+    {
+        private readonly string DevicePrefix = $"E2E_{nameof(BulkOperationsE2ETests)}_";
+        private readonly string ModulePrefix = $"E2E_{nameof(BulkOperationsE2ETests)}_";
+
+        private static TestLogging _log;
+        private readonly ConsoleEventListener _listener;
+
+        public BulkOperationsE2ETests()
+        {
+            _listener = TestConfig.StartEventListener();
+            _log = TestLogging.GetInstance();
+        }
+
+        [TestMethod]
+        public async Task BulkOperations_UpdateTwins2_Device()
+        {
+            await BulkOperations_UpdateTwins2(false).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task BulkOperations_UpdateTwins2_Module()
+        {
+            await BulkOperations_UpdateTwins2(true).ConfigureAwait(false);
+        }
+
+        private async Task BulkOperations_UpdateTwins2(bool setModule)
+        {
+            var tagName = Guid.NewGuid().ToString();
+            var tagValue = Guid.NewGuid().ToString();
+
+            TestDevice testDevice = setModule ? null : await TestDevice.GetTestDeviceAsync(DevicePrefix).ConfigureAwait(false);
```

Comment:
This code is very hard to read. Instead of switching between testDevice and testModule, it would have been clearer to write two separate tests.